### PR TITLE
Evaluator of configuration policies written in Rego files

### PR DESCRIPTION
**Pull request by shlomitk1**
Signed-off-by: Shlomit Koyfman <shlomitk@il.ibm.com>

Will close https://github.com/fybrik/fybrik/issues/794
Implementation of OPA-based evaluation of config policies according to #932

Initial approach: the policies are added in the deployment time. Changes of policies and infrastructure are not tracked.
This PR also changes the way data path is constructed, removing hard-coded capabilities, like "read" and "copy"

### CODE REVIEW

**Review by simanadler — CHANGES_REQUESTED** [review]
I didn't read through all of the code yet, but there are some important initial comments.

File: pkg/adminconfig/rego_file_evaluator.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// Type definitions for parsing OPA response
+// A list of decisions per capability, e.g. {"read": {"deploy": true}, "write": {"deploy": false}}
+type RuleDecisionList []DecisionPerCapabilityMap
+
+// A directory containing rego files that define admin config policies
+const RegoPolicyDirectory = "/tmp/adminconfig/"
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/adminconfig/rego_file_evaluator.go:28]: This directory inside the pod is set during the deployment, and the files are copied automatically from a configmap. Similar to what we do with taxonomy files or governance policies.

File: manager/controllers/app/plotter_generator.go [hunk 2]
```diff
@@ -220,8 +220,8 @@ func (p *PlotterGenerator) AddFlowInfoForAsset(item DataInfo, appContext *app.Fy
 			readAssetID = datasetID + "-copy"
 		}
 		actions := createActionStructure(readSelector.Actions)
-		if len(item.Configuration.ConfigDecisions[app.Read].Clusters) > 0 {
-			readCluster = item.Configuration.ConfigDecisions[app.Read].Clusters[0]
+		if len(item.Configuration.ConfigDecisions[app.Read].DeploymentRestrictions.Clusters) > 0 {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/plotter_generator.go:223]: I understand what you are saying. That it does not matter whether the capability is called read, write or xyz as long as it has a required source, sink or api, and that's how we connect them. OK, I'll try to change.

File: charts/fybrik/files/adminconfig/default_policies.rego [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+# write capability deployment
+config[{"write": decision}] {
+    write_request := input.request.usage.write 
+    policy := {"ID": "write-default", "description":"Write capability is requested for workloads that write data"}
+    decision := {"policy": policy, "deploy": write_request}
+}
+
+# allow implicit copies by default
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/default_policies.rego]: Moved to the quick-start policies.

File: charts/fybrik/files/adminconfig/quickstart_policies.rego [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+# configure where the read capability will be deployed
+config[{"read": decision}] {
+    input.request.usage.read == true
+    policy := {"ID": "read-location", "description":"Deploy read in the workload cluster"}
+    decision := {"policy": policy, "restrictions": {"clusters": [ input.workload.cluster.name]}}
+}
+
+# copy requested by the user
+config[{"copy": decision}] {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/quickstart_policies.rego:25]: OK.

File: charts/fybrik/files/adminconfig/quickstart_policies.rego [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package adminconfig
+
+# configure where transformations take place
+config[{"transform": decision}] {
+    policy := {"ID": "transform-geo", "description":"Governance based transformations must take place in the geography where the data is stored"}
+    clusters := [ data.clusters[i].name | data.clusters[i].metadata.region == input.request.dataset.geography ]
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/quickstart_policies.rego:6]: @simanadler In this case, the evaluator does not require access to infrastructure, correct?

File: charts/fybrik/files/adminconfig/quickstart_policies.rego [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+# configure when implicit copies should be made
+config[{"copy": decision}] {
+    input.request.usage.read == true
+    input.request.dataset.geography != input.workload.cluster.metadata.region
+    count(input.actions) > 0
+    clusters :=  [ data.clusters[i].name | data.clusters[i].metadata.region == input.request.dataset.geography ]
+    policy := {"ID": "copy-remote", "description":"Implicit copies should be used if the data is in a different region than the compute, and transformations are required"}
+    decision := {"policy": policy, "deploy": true, "restrictions": {"clusters": clusters}}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/quickstart_policies.rego:39]: Done

File: charts/fybrik/files/adminconfig/quickstart_policies.rego (hunk 5 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/quickstart_policies.rego]: Done

File: charts/fybrik/files/adminconfig/default_policies.rego [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package adminconfig
+
+# out-of-the-box policies
+
+# read capability deployment
+config[{"read": decision}] {
+    read_request := input.request.usage.read
+    policy := {"ID": "read-default", "description":"Read capability is requested for read workloads"}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/files/adminconfig/default_policies.rego:8]: I'll add in a separate PR.

File: manager/controllers/app/plotter_generator.go [hunk 8]
```diff
@@ -159,8 +159,8 @@ func (p *PlotterGenerator) AddFlowInfoForAsset(item DataInfo, appContext *app.Fy
 		}
 		var copyDataAssetID = datasetID + "-copy"
 		actions := createActionStructure(copySelector.Actions)
-		if len(item.Configuration.ConfigDecisions[app.Copy].Clusters) > 0 {
-			copyCluster = item.Configuration.ConfigDecisions[app.Copy].Clusters[0]
+		if len(item.Configuration.ConfigDecisions[app.Copy].DeploymentRestrictions.Clusters) > 0 {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/plotter_generator.go]: Fixed.

File: manager/controllers/app/plotter_generator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/plotter_generator.go]: Fixed.

File: manager/controllers/app/plotter_generator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/plotter_generator.go]: Removed hard-coded logic. Storage allocation is still an issue, I'll open a discussion.

**Pull request closed without merging**